server/models/visit.js

```diff
@@ -0,0 +1,157 @@
+    // do a little transformin' and this might barf, so do it before `done`
+
+    // if nothing was found, continue
+    if (!results) { return; }
+
+    // this is to work around our laziness in SELECT * above
+    results.id = results.visitId;
+    delete results.visitId;
+
+    // keep visit properties top level, move all other page properties under a userPage key
+    var newResult = {};
+    ['id', 'url', 'title', 'visitedAt'].forEach(function(item) {
+      newResult[item] = results[item];
+      delete results[item];
+    });
+    newResult.userPage = results;
+    return newResult;
+  },
+  // 1. find a complex way to combine both queries on the DB side
+  // 2. (simpler) perform two simple queries and roll them together here
+  get: function(userId, visitId, cb) {
+    var name = 'models.visit.get';
+    _verbose(name + ' called', userId, visitId);
+    // it's actually way simpler to SELECT *, and re-select additional columns,
+    // vs enumerating everything just to avoid two 'id's in the results.
+    var query = 'SELECT visits.id as visit_id, visits.user_id as user_id, * ' +
+    'FROM visits LEFT JOIN user_pages ON visits.user_page_id = user_pages.id ' +
+    'WHERE visits.id = $1 AND visits.user_id = $2';
+    var params = [visitId, userId];
+    postgres.query(query, params)
+      .then(function(results) {
+        // return a promise that resolves to the transformed results
+        return q(visit._transform(results));
+      })
+      .done(visit._onFulfilled.bind(visit, name + ' succeeded', cb),
+            visit._onRejected.bind(visit, name + ' failed', cb));
+  },
+  create: function(userId, visitId, visitedAt, url, urlHash, title, cb) {
+    // XXX we create only a few user_page fields synchronously; the rest
+    // are filled in async by the scraper worker
+    var name = 'models.visit.create';
+    _verbose(name + ' called', userId, visitId, visitedAt, url, title);
+    // create the user_page if it doesn't exist, and return the user_page id
+    // whether or not you just created it
+    var lazyCreateUserPageQuery =
+      'WITH new_page AS (  ' +
+      '  INSERT INTO user_pages (id, user_id, url, raw_url, url_hash, title, created_at, updated_at) ' +
+      '  SELECT $1, $2, $3, $3, $4, $5, $6, $6 ' +
+      '  WHERE NOT EXISTS (SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4) ' +
+      '  RETURNING id ' +
+      ') SELECT id FROM new_page ' +
+      'UNION SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4';
+    var lazyCreateParams = [uuid.v4(), userId, url, urlHash, title, new Date().toJSON()];
+    var createVisitQuery = 'INSERT INTO visits ' +
+      '(id, user_id, user_page_id, visited_at, updated_at) ' +
+      'VALUES ($1, $2, $3, $4, $4)';
+    var userPageId;
+
+    // try to insert into pg; handle pg errors; insert into es; then we're done.
+    // don't pass results to fulfillment callback, because they aren't used by the caller.
```

Comment:
actually, if create returns the record, then the worker can verify that the returned results look sane (or not). something to consider.